---
title: "A General Method to map Single-cell Probabilistic Trait Loci of the Genome (Simulated Dataset)"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{A General Method to map Single-cell Probabilistic Trait Loci of the Genome (Simulated Dataset)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



## Introduction

### Dataset

This report presents the analysis of a simulated dataset. 

The dataset is obtain from a simulator. The implemented model describes the behaviour of an auto-regulated single gene.
The production of protein is control by the following equation:
\[P = \frac{\alpha_0 + \alpha_1 (\frac{P}{K})^h}{1+(\frac{P}{K})^h}\]

We use two sets of parameter to produce cell populations:

\[\left\lbrace
  \begin{array}{l}
    \alpha_0 = 0.1 \\
    \alpha_1 = 10 \\
    K = 1.6
  \end{array}
\right.\]

\[\left\lbrace
  \begin{array}{l}
    \alpha_0 = 6.3\\
    \alpha_1 = 12\\
    K = 10
  \end{array}
\right.\]

For each individual we generate 20000 cells according to one of this two set of parameters and a given level of error on the parameter values.

The dataset is freely available as an R package called `ptldata` (https://github.com/fchuffar/ptldata).

```{r}
library(ptldata)
```

The R package called `ptldata` exposes following objects:

#### indiv

`indiv` describes the simulated individuals. It is a data frame that contains the description of the 1490
 individuals. It is compose dof  1490 rows and 8 variables:

  * `err`, the value of the error introducted in the paramater selection level.
  * `all`, the discriminant parental allele orign.
  * `a0`, the a0 parameter value.
  * `a1`, rhe a1 parameter value.
  * `k`, the k parameter value.
  * `m1`, the first moment of the single cell distribution.
  * `m2`, the second moment of the single cell distribution.
  * `m3`, the third moment of the single cell distribution.

In this study, we only focus on individuals that have been generated with the level of error of 0.2 and, for computanial reason we down sample 80 individuals.

```{r}
err = 0.2
nb_indiv = 80
idx = sample(which(indiv$err==err), nb_indiv)
cells = cells[idx]
indiv = indiv[idx, ]
genodata = genodata[,idx]
```

This picture shows the distribution of the first, second and third moment of the single cells data associated to each individual in grey and its composition (in blue and red) if we considere discriminant parental allele orign. Red and blue histograms of the first and second moment are overlaping. The two same histograms of the third moemnt are separated.

```{r,fig.height=3, fig.width=9}
layout(matrix(1:3, 1), respect=TRUE)
for (m in c("m1","m2","m3")){  
  h = hist(indiv[[m]], nclass=20, 
    main=paste("moment", m, sep=""), xlab="", ylab="", 
    col=adjustcolor("grey", alpha=0.3))
  for (all in unique(indiv$all)) {
    h = hist(indiv[indiv$all==all,][[m]], plot=FALSE, breaks=h$breaks)    
    lines(h, col=adjustcolor(all+1, alpha=0.3))
  }
}
```

#### genodata

`genodata` specify the genotype of the simulated individuals. It is 
a data frame with 210 rows and 1490 columns whose rows are markers and columns are individuals. The
marker 101 corresponds to the 'all' variable of the 'indiv' data
frame, it is the parental allele linked with the phenotype. From
this marker 101, genotype is propagated (upstream and downstream)
with recombination factor of 5

The following image figures out the parental origin of segregant genomes according to marker.

```{r,fig.height=6, fig.width=6}
tetas = compute_teta(genodata)
matlab::imagesc(t(as.matrix(genodata)), 
  main="Parental origin of segregant genomes", 
  xlab="markers", ylab="segrgants")
```

The following plot represents the distribution of the recombination factor (teta) between marker.

```{r,fig.height=6, fig.width=6}
plot(density(tetas), 
  main=paste("Recombination Fraction Distribution"), 
  xlab=paste("mean(teta)=", signif(mean(tetas),3), sep=""))
abline(v=mean(tetas), lty=2)
```

#### cells

`cells` emdeb simulated single cell values (phenotypes) of individuals. 
It is a list of the 20000 single cell values (phenotype) of the 1490
individuals described by the variable 'indiv'. It is composed of 1490 items,
each item is a vector of 20000 integer

```{r,fig.width=6, fig.height=6}
# plot_dist(ptl_mapping_result, main=main, col=col)
library(ptlmapper)
```

### Method

The main purpose of this analysis is to presents a general method allowing to map Probabilistic Trait Loci (PTL).
This analysis is composed of the following steps:

#### analysis of the single cell signal (phenotype)

From the list of single-cell values we build histograms. 
We obtain a list of histogram, one per individual. 
All this histograms have the same `breaks` values. 

```{r}
pheno_hists = build_pheno_hists(cells, bin_width=1)
```

#### computation of the Kantorovitch distance matrix

From the list of histograms we build the Kantorovitch distance matrix. 
Since Kantorovitch distance between two individuals is computed using their histograms, 
it is important that the `breaks` values.

```{r}
kd_matrix = build_kd_matrix(pheno_hists)
```

#### computation of the multivariate moment matrix

From the list of single values stored in the histograms we build the multivariate moment matrix. 

```{r}
mm_matrix = build_mmoments_matrix(pheno_hists)
```

#### preparation of the genotype

The methods needs an annotated genome to work fine. Here, we four colums to our genodata data frame: 

  * `prob_name`, the probe name associated to the marker.
  * `chromosome`, the chromosome on which the marker is.
  * `position`, the position of the marker in the chromosome.
  * `rec_fractions`, the recombination factor associated to the marker.

In `genodata`, "2" stands for undetermined and is replaced by NA.

```{r}
bckg = names(genodata)
genodata$chromosome = rep(1, nrow(genodata))
genodata$position = 1:nrow(genodata)
genodata$prob_name = paste("marker", 1:nrow(genodata), sep="_")
genodata$rec_fractions = 0.05

genodata_ptl = preprocess_genodata(genodata, bckg)
```


#### do the ptl mapping using Kantovitch method

```{r}
kanto_analysis = ptl_scan(kd_matrix, genodata_ptl, nb_perm=3)
```

#### do the ptl mapping using multivariate moment method

```{r}
mmoments_analysis = ptl_scan(mm_matrix, genodata_ptl, nb_perm=3, method="mmoment")
```


#### do the qtl mapping using the classic method

```{r}
```


### Result

```{r}
```




```{r,fig.height=6, fig.width=12}


###################################################
### code chunk number 7: simdata_pres.Rnw:553-556
###################################################

main=paste("err=", err, sep="")
# ptl_mapping_filename = paste("cache/simdata", err, nb_perm, "ptl_mapping.Rds", sep="_")
# genodata = g
# cells = c
# bckg = bckg
# nb_perm=nb_perm
# nb_dim=0
# bin_width=1
# ptl_mapping_filename=ptl_mapping_filename
# DO_MMOMENTS=TRUE
# errs = c(0.05, 0.01, 0.005)
# minimal_proportion=0.1
# COMPUTE_KD_MATRIX=TRUE
# DO_KANTO=TRUE
# DO_RQTL=TRUE
# nb_moments=3
# LIGHTWEIGHT=FALSE
ptl_mapping_result = ptl_mapping(genodata, cells, bckg, nb_perm=3, nb_dim=0, bin_width=1, 
# ptl_mapping_filename=ptl_mapping_filename, 
DO_MMOMENTS=TRUE)

layout(matrix(1:8, 2, byrow=TRUE), respect=TRUE)
plot_rqtl(ptl_mapping_result, main=paste(main, dim(genodata)[1], "markers and", dim(genodata)[2], "indiviuals"), which_pheno=1)
plot_rqtl(ptl_mapping_result, main=main, which_pheno=2)
marker_names = rownames(get_best_markers_rptl(ptl_mapping_result))
col = mn_2_col(ptl_mapping_result, marker_names[1])
plot_mds(ptl_mapping_result, main=main, col=col)
plot_wilks(ptl_mapping_result, main=main, method="kanto")
plot_wilks(ptl_mapping_result=ptl_mapping_result, pa=ptl_mapping_result$mmoments_analysis, main=paste(nb_indiv, "indiv (mm)"), method="mmoment")
plot_can(ptl_mapping_result, marker_names[1], main=main, col=col)
# plot_noise(ptl_mapping_result, main=main, col=col)
plot_dist(ptl_mapping_result, main=main, col=col)
plot_empirical_test(ptl_mapping_result,main=main)
```


## Aknowledgements